source/lambda/helper/python/kendraHelper.py

```diff
@@ -25,11 +25,78 @@ def indexDocument(self,
+        s3helper = S3Helper()
+        policyData = None
+        
+        try:
+            policyData = s3helper.readFromS3(s3bucket,
+                                             policyFilepath,
+                                             os.environ['AWS_REGION'])
+    
+        # the normal case of a file not provided is handled.  If any other error
+        # occur the indexing will proceed without the membership tags in the policy file
+        except ClientError as e:
+            policyData = None
+            # NoSuchKey is the expected exception, any other means an error
+            if e.response['Error']['Code'] == 'NoSuchKey':
+                print("no kendra policy file found, only default membership will be applied")
+            else:
+                print("ClientError exception from s3helper.readFromS3: " + str(e))
+                    
+        # an error that should be investigated
+        except Exception as e:
+            policyData = None
+            print("unspecified exception from s3helper.readFromS3: " + str(e))
+        
+        # accessControlList will contain the default persona membership, the function call provided
+        # tags if different from default, and any additonal membership tags in the metadata policy
+        # json file was given in s3 with the document
+        accessControlList = []
+        
+        # the default membership for all documents
+        defaultMembership = {}
+        defaultMembership['Name'] = 'everybody'
+        defaultMembership['Type'] = 'GROUP'
+        defaultMembership['Access'] = 'ALLOW'
+        accessControlList.append(defaultMembership)
+        
+        # if a different membership tag was provided in the function call, add it
+        # as well
+        if tag != 'everybody':
+            tagMembership['Name'] = tag
+            tagMembership['Type'] = 'GROUP'
+            tagMembership['Access'] = 'ALLOW'
+            accessControlList.append(tagMembership)
+        
+        # if the policy file exists, it may contain additional membership tags.  Parsing
+        # error may happen and will be caught
+        try:
+            if policyData != None:
+                
+                policyAccessList = json.loads(policyData)
+                
+                for membership in policyAccessList['AccessControlList']:
+                    
+                    # no need for tags in the policy that may have been already added above
+                    if membership['Name'] != 'everybody' and membership['Name'] != tag:
+                        accessControlList.append(membership)
+        
+        # indexing will proceed without the membership tags in the policy file
+        except Exception as e:
+            print("exception while processing policy file " + policyFilepath + str(e))
+    
+        print("document will have the following membership policy in Kendra: " + json.dumps(accessControlList))
```

Comment:
could also add document id?

'Document {} will have the following membership policy in Kendra:{}'.format(documentId, json.dumps(accessControlList))